### ISSUE TITLE
How does runtime subscription work?

### ISSUE THREAD

**Issue by daimeng** [mobxjs/mobx#248]
Hi guys. A little background. Currently working as part of a small group to completely overhaul a backbone frontend. I've chosen MobX recently as our primary state management candidate to use with React. We have a large team to expand this to later potentially, so I needed something easy to ramp up. While I find the overall concepts easy to follow, there are some parts that still seem like black magic to me.

I want to get around to reading through the source for this library later, but I'm completely bogged down. My biggest question is how runtime subscription works. Maybe I'm missing something obvious, but how does say autorun(() => console.log(observableVar)) know what variables were used inside its function argument?

**Comment by AriaFallah** [comment]
@daimeng 

MobX tries to act as much like a proxy sitting in front of an array as it can without being able to use ES6 proxies. MobX arrays aren't actually native JavaScript arrays. Internally they use a JS array, but all of the array prototype methods are overwritten to either `reportObserved` or `reportChanged` before actually running calling the method on the internal collection. With MobX objects it's similar except they're just regular objects filled with observable values.

Knowing this, autorun will register your function to run again whenever `reportObserved()` is called, and if it isn't called, that autorun won't run again.

To give a more concrete example, if you had:

```js
const x = observable([])
autorun(() => console.log(x))

x.push(1)
x.push(2)
x.push(3)
```

the autorun would only run once. MobX would run `console.log(x)`, realized that nothing inside that calls `reportObserved`, and will never run again.

So instead you think to yourself, how about I log `x[0]` because I know `x[someValue]` calls `reportObserved()`:

```js
const x = observable([])
autorun(() => console.log(x[0]))

x.push(1)
x.push(2)
x.push(3)
```

autorun will still only run once because that's only true if `x[someValue]` isn't undefined.

---

So how do you get around this? With arrays it's as simple as calling `slice()`

```js
const x = observable([])
autorun(() => console.log(x.slice()))

x.push(1)
x.push(2)
x.push(3)
```
and

```js
const x = observable([])
autorun(() => console.log(x.slice()[0]))

x.push(1)
x.push(2)
x.push(3)
```
for the respective problems.

---

Likewise

With plain observable values to log it'd be

```js
const x = observable(1)
autorun(() => console.log(x.get()))
x.set(1)
x.set(2)
x.set(3)
```

and with MobX maps to log the whole thing it'd be

```js
const x = map({})
autorun(() => console.log(x.toJs()))
x.set(1)
x.set(2)
x.set(3)
```
---

The gist of the idea here is that you need to find a way to trigger `reportObserved` the easiest of which is just by calling the respective defensive copying methods of each observable type.

**Comment by mweststrate** [comment]
@daimeng in two lines: `autorun` and `computed` put the function the execute on a stack. Whenever an observable values is accessed, MobX associates that value with the function that was started by `autorun` / `computed`. Now if that value changes in the future, the associated `autorun` / `computeds will be retriggered. It is explained in greater depth [here](https://medium.com/@mweststrate/becoming-fully-reactive-an-in-depth-explanation-of-mobservable-55995262a254).

[offtopic]
@ariafalah yes `computed(expr).observe(effect)` might make it a bit more clear. (autorun is actually `computed(expr).observe(noop)`). I'm thinking about introducing this as explicit concept in the api, but I'm not sure how to name it yet. As overload for `autorun`, or `observe`, or under a separate name (`react(ion)`)? Note that this pattern might be confusing in another way, if expr just returns an object, but the individual members are only accessed in the `effect`, those members wouldn't be tracked. With `autorun` you cannot make thát mistake. So still doubting what is the best thing to recommend. Note that in MobX 2.2 `autorun(() => someAction(expr))`  would also have the same effect as `computed(expr).observe(effect)` if `const someAction = action(data => effect)`. So there are many ways to express the same thing, just looking for what is the most clear :)

**Comment by daimeng** [comment]
@AriaFallah Thanks for the explanations. I've been playing around with some of your examples in jsfiddle and how observable works with different kinds of values is becoming a lot more clear to me now. This is a really useful reference to keep.

@mweststrate Yes I read that article! It was a large part of what convinced me to try MobX. At the time I did understand in some capacity that the runtime subscription worked by a proxy-ish behavior. But while reading the article, I had assumed (stupidly) that autorun function was not run once initially. I thought somehow the values were being discovered in another way for autorun. Of course, this made no sense and I found that out with further testing.
